sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -233,3 +236,31 @@ private[csv] class UnivocityParser(
     }
   }
 }
+
+private[csv] object UnivocityParser {
+  def tokenizeStream(
+      inputStream: InputStream,
+      header: Boolean,
+      settings: CsvParserSettings): Iterator[Array[String]] = new Iterator[Array[String]] {
+    private val parser = new CsvParser(settings)
+    // Note that, here we assume `inputStream` is the whole file that might include the header.
+    parser.beginParsing(inputStream)
+    private var nextRecord = {
+      if (header) {
+        parser.parseNext()
```

Comment:
Yup, existing code path processes line by line so comments and empty lines had to manually skipped because it threw an error if each line has only commented line or empty. This is nicely handled in Univocity and the initial version of CSV actually did this in this way.